helix-term/src/handlers/completion.rs

```diff
@@ -291,6 +300,213 @@ fn request_completion(
     });
 }
 
+fn path_completion(
+    cursor: usize,
+    text: helix_core::Rope,
+    doc: &helix_view::Document,
+) -> Option<BoxFuture<'static, anyhow::Result<Vec<CompletionItem>>>> {
+    if !doc.supports_path_completion() {
+        return None;
+    }
+
+    // TODO find a good regex for most use cases (especially Windows, which is not yet covered...)
+    // currently only one path match per line is possible in unix
+    static PATH_REGEX: Lazy<rope::Regex> =
+        Lazy::new(|| rope::Regex::new(r"((?:~|\$HOME|\$\{HOME\})?(?:\.{0,2}/)+.*)").unwrap());
+
+    let cur_line = text.char_to_line(cursor);
+    let start = text.line_to_char(cur_line).max(cursor.saturating_sub(1000));
+    let line_until_cursor = text.slice(..).regex_input_at(start..cursor);
+
+    let (dir_path, typed_file_name) = PATH_REGEX
+        .find_iter(line_until_cursor)
+        .last() // this is a workaround for missing "match end of slice" support in `rope::Regex`
+        .and_then(|matched_path| {
+            let end = text.byte_to_char(matched_path.end());
+            if end != cursor {
+                return None;
+            }
+            let start = text.byte_to_char(matched_path.start());
+            let matched_path = &text.slice(start..end).to_string();
+
+            // resolve home dir (~/, $HOME/, ${HOME}/) on unix
+            #[cfg(unix)]
+            let mut path = {
+                static HOME_DIR: Lazy<Option<std::ffi::OsString>> =
+                    Lazy::new(|| std::env::var_os("HOME"));
+
+                let home_resolved_path = if let Some(home) = &*HOME_DIR {
+                    let first_separator_after_home = if matched_path.starts_with("~/") {
+                        Some(1)
+                    } else if matched_path.starts_with("$HOME") {
+                        Some(5)
+                    } else if matched_path.starts_with("${HOME}") {
```

Comment:
You mean implement our own (more general) kind of shell expansion in `helix-stdx`? What kind of dependency are you talking about?

Reply:
I meant https://docs.rs/shellexpand/latest/shellexpand/

A lot of complexity in this crate comes from the fact that in the past converting raw OsString <-> bytes was ugly but these days it's almost trivial (since we can use a regex).

Basically just iterate all matches for a regex `\$\{[\w@/\\.\-+#$%~=:]+\}||\$\w+` and try to replace them with the environment variable. I had a function written up. If you feel uncomfortable with that I can also write it

Reply:
I mean I could certainly dive into this, but it seems you already have a pretty good idea how to do all of that, so you're likely faster (including much more research regarding the path recognition regex).

Maybe it makes sense for you to create a PR with `gf` + path regex + shell expansion, and I base this path-completion PR on top of that?

I can help reviewing, if that helps getting this merged faster.

Reply:
I would prefer doing a PR against your branch/fork

Reply:
That's good for me as well.

Reply:
This was solved with https://github.com/Philipp-M/helix/pull/4